src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1112,6 +1112,109 @@ protected static ImmutableArray<string> CheckTupleElementNames(int cardinality,
         public bool HasImplicitConversion(ITypeSymbol fromType, ITypeSymbol toType)
             => fromType != null && toType != null && this.ClassifyCommonConversion(fromType, toType).IsImplicit;
 
+        /// <summary>
+        /// Checks if <paramref name="symbol"/> is accessible from within <paramref name="within"/>. An optional qualifier of type
+        /// <paramref name="throughType"/> is used to resolve protected access for instance members. All symbols are
+        /// required to be from this compilation or some assembly referenced by this compilation. <paramref name="within"/> is required
+        /// to be an <see cref="ITypeSymbol"/> or <see cref="IAssemblySymbol"/>.
+        /// </summary>
+        public bool IsSymbolAccessibleWithin(
+            ISymbol symbol,
+            ISymbol within,
+            ITypeSymbol throughType = null)
+        {
+            if (symbol is null)
+            {
+                throw new ArgumentNullException(nameof(symbol));
+            }
+
+            if (within is null)
+            {
+                throw new ArgumentNullException(nameof(within));
+            }
+
+            if (!(within is INamedTypeSymbol || within is IAssemblySymbol))
+            {
+                throw new ArgumentException(CodeAnalysisResources.IsSymbolAccessibleBadWithin, nameof(within));
+            }
+
+            checkInCompilation(symbol, nameof(symbol));
+            checkInCompilation(within, nameof(within));
+            if (!(throughType is null))
+            {
+                checkInCompilation(throughType, nameof(throughType));
+            }
+
+            return IsSymbolAccessibleWithinCore(symbol, within, throughType);
+
+            void checkInCompilation(ISymbol s, string parameterName)
+            {
+                var containingAssembly = computeContainingAssembly(s);
+                if (!assemblyIsOk(containingAssembly))
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.IsSymbolAccessibleWrongAssembly, parameterName), parameterName);
+                }
+            }
+
+            bool assemblyIsOk(IAssemblySymbol a)
+            {
+                if (a == this.Assembly)
+                {
+                    return true;
+                }
+
+                foreach (var reference in this.References)
+                {
+                    var moduleOrAssembly = this.CommonGetAssemblyOrModuleSymbol(reference);
```

Comment:
>this.CommonGetAssemblyOrModuleSymbol(reference); [](start = 43, length = 48)

It looks like this API can return null and the line below would crash in that case. See Microsoft.CodeAnalysis.CSharp.UnitTests.ReferenceManagerTests.DuplicateReferences for example. Please add a test covering a scenario like that.